modules/core/04-channel/keeper/handshake.go

```diff
@@ -152,72 +152,55 @@ func (k Keeper) ChanOpenTry(
 		)
 	}
 
-	// handle multihop case
+	var counterpartyHops []string
 	if len(connectionHops) > 1 {
-		kvGenerator := func(mProof *types.MsgMultihopProofs, lastHopConnectionEnd *connectiontypes.ConnectionEnd) (string, []byte, error) {
-			// check version support
-			if err := checkVersion(lastHopConnectionEnd, order); err != nil {
-				return "", nil, err
-			}
-
-			// channel end storekey of the counterparty channel on the other end of the multihop channel
-			key := host.ChannelPath(counterparty.PortId, counterparty.ChannelId)
-
-			// connection hops
-			counterpartyHops, err := mProof.GetCounterpartyConnectionHops(k.cdc, &connectionEnd)
-			if err != nil {
-				return "", nil, err
-			}
-
-			// expectedCounterparty is the counterparty of the counterparty's channel end (i.e self)
-			expectedCounterparty := types.NewCounterparty(portID, "")
-			expectedChannel := types.NewChannel(types.INIT, order, expectedCounterparty, counterpartyHops, counterpartyVersion)
-
-			// expected value bytes
-			bz, err := k.cdc.Marshal(&expectedChannel)
-			if err != nil {
-				return "", nil, err
-			}
-
-			return key, bz, nil
+		var multihopProof types.MsgMultihopProofs
+		if err := k.cdc.Unmarshal(proofInit, &multihopProof); err != nil {
+			return "", nil, err
 		}
 
-		if err := k.connectionKeeper.VerifyMultihopMembership(
-			ctx, connectionEnd, proofHeight, proofInit,
-			connectionHops, kvGenerator); err != nil {
+		// get the last hop connection on the other side of the multihop channel
+		// the last hop connection is the connection end on the chain before the counterparty multihop chain
+		lastHopConnectionEnd, err := multihopProof.GetLastHopConnectionEnd(k.cdc, connectionEnd)
```

Comment:
There is still some leakage of the multi hop abstraction at 04-channel, but I couldn't see any way around it. Open to ideas.

Reply:
I took a deeper look at this today and I think I see a path forward, but it is annoying due to import cycles. I am working on a branch to merge multi-hop and single hop logic into the existing VerifyXXX() functions by changing the single hop proof into a multi-hop proof. The multi-hop proof logic then degenerates into standard single hop proof logic.

However, I need to to some moderate refactoring to get this approach to work. I'll share a draft PR so you can see the approach hopefully next week. The main changes are to:
- move multihop message definitions under commitment types (at least not under channel/tx types)
- remove dependency on tmclient from the multihop verification code (need to break import cycle and shouldn't really be hardcoded to tm anyway...)
- need to add connectionHops as arg to connection keeper interface
- add a multihop membership verification method to client state interface and implement for tendermint

Hopefully I can do this in a way where I can pick just one verification function and show how it would look. This would also remove the key generator functions as well.